File: TouchNothing/Services/SupabaseService.swift
import Foundation
import Supabase

enum SupabaseServiceError: Error, Equatable {
    case nicknameTaken
    case invalidCredentials
    case accountLocked
    case invalidPin
    case invalidNickname
    case invalidSessionToken
    case nicknameNotRegistered
    case serverUpgradeRequired
    case notConfigured
    case networkFailure
}

final class SupabaseService {
    static let shared = SupabaseService()

    private var client: SupabaseClient?

    private init() {
        configureClientIfPossible()
    }

    private func configureClientIfPossible() {
        guard let configuration = try? SupabaseConfig.load() else { return }
        client = SupabaseClient(supabaseURL: configuration.url, supabaseKey: configuration.anonKey)
    }

    var isConfigured: Bool {
        client != nil
    }

    func registerUser(nickname: String, pin: String) async throws -> AuthSession {
        try await fetchAuthSession(
            "register_user",
            params: NicknameParams(pNickname: nickname, pPin: pin)
        )
    }

    func loginUser(nickname: String, pin: String) async throws -> AuthSession {
        try await fetchAuthSession(
            "login_user",
            params: NicknameParams(pNickname: nickname, pPin: pin)
        )
    }

    func submitSession(
        nickname: String,
        durationSeconds: Int,
        sessionToken: String
    ) async throws {
        guard durationSeconds >= AppConstants.minSessionSeconds else { return }
        try await executeVoidRPC(
            "submit_session",
            params: SessionParams(
                pNickname: nickname,
                pDuration: durationSeconds,
                pSessionToken: sessionToken
            )
        )
    }

    func fetchUserStats(nickname: String) async throws -> UserStats {
        guard let client else { throw SupabaseServiceError.notConfigured }

        let normalizedNickname = NicknameValidator.normalized(nickname)

        do {
            let stats: [UserStats] = try await fetchRPC(
                "get_user_stats",
                params: UserStatsParams(pNickname: normalizedNickname)
            )
            if let statsRow = stats.first {
                return statsRow
            }
        } catch {
            // Fall back to direct table read if RPC is unavailable.
        }

        let stats: [UserStats] = try await client
            .from("user_stats")
            .select("total_duration_seconds, session_count")
            .ilike("nickname", pattern: normalizedNickname)
            .limit(1)
            .execute()
            .value

        guard let statsRow = stats.first else {
            throw SupabaseServiceError.networkFailure
        }

        return statsRow
    }

    func fetchLeaderboard(limit: Int = AppConstants.leaderboardLimit) async throws -> [LeaderboardEntry] {
        guard let client else { throw SupabaseServiceError.notConfigured }

        let entries: [LeaderboardEntry] = try await client
            .from("records")
            .select()
            .order("duration_seconds", ascending: false)
            .limit(limit)
            .execute()
            .value

        return entries
    }

    func fetchTotalLeaderboard(limit: Int = AppConstants.leaderboardLimit) async throws -> [TotalLeaderboardEntry] {
        guard let client else { throw SupabaseServiceError.notConfigured }

        let entries: [TotalLeaderboardEntry] = try await client
            .from("user_stats")
            .select("nickname, total_duration_seconds")
            .gt("total_duration_seconds", value: 0)
            .order("total_duration_seconds", ascending: false)
            .limit(limit)
            .execute()
            .value

        return entries
    }

    private func fetchAuthSession(
        _ function: String,
        params: NicknameParams
    ) async throws -> AuthSession {
        guard let client else { throw SupabaseServiceError.notConfigured }

        do {
            let response = try await client.rpc(function, params: params).execute()
            if let session = try? JSONDecoder().decode(AuthSession.self, from: response.data) {
                return session
            }
            if let token = try? JSONDecoder().decode(String.self, from: response.data) {
                return AuthSession(
                    nickname: NicknameValidator.normalized(params.pNickname),
                    sessionToken: token
                )
            }
            throw SupabaseServiceError.serverUpgradeRequired
        } catch let postgrestError as PostgrestError {
            throw SupabaseRPCErrorMapper.map(postgrestError)
        } catch let serviceError as SupabaseServiceError {
            throw serviceError
        } catch {
            if SupabaseRPCErrorMapper.isLikelyEmptyAuthResponse(error) {
                throw SupabaseServiceError.serverUpgradeRequired
            }
            throw SupabaseRPCErrorMapper.map(error)
        }
    }

    private func executeVoidRPC(
        _ function: String,
        params: some Encodable,
        mapErrors: Bool = true
    ) async throws {
        guard let client else { throw SupabaseServiceError.notConfigured }

        do {
            try await client.rpc(function, params: params).execute()
        } catch {
            throw rpcError(from: error, mapErrors: mapErrors)
        }
    }

    private func fetchRPC<T: Decodable>(
        _ function: String,
        params: some Encodable,
        mapErrors: Bool = true
    ) async throws -> T {
        guard let client else { throw SupabaseServiceError.notConfigured }

        do {
            return try await client
                .rpc(function, params: params)
                .execute()
                .value
        } catch {
            throw rpcError(from: error, mapErrors: mapErrors)
        }
    }

    private func rpcError(from error: Error, mapErrors: Bool) -> SupabaseServiceError {
        if mapErrors {
            return SupabaseRPCErrorMapper.map(error)
        }
        return .networkFailure
    }
}

File: app/src/main/java/com/touchnothing/app/data/remote/SupabaseService.kt
package com.touchnothing.app.data.remote

import com.touchnothing.app.data.model.AuthSession
import com.touchnothing.app.data.model.LeaderboardEntry
import com.touchnothing.app.data.model.NicknameParams
import com.touchnothing.app.data.model.SessionParams
import com.touchnothing.app.data.model.TotalLeaderboardEntry
import com.touchnothing.app.data.model.UserStats
import com.touchnothing.app.data.model.UserStatsParams
import com.touchnothing.app.util.AppConstants
import com.touchnothing.app.util.NicknameValidator
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.isSuccess
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json

class SupabaseService {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val client: HttpClient? = if (SupabaseConfig.isConfigured) {
        HttpClient(OkHttp) {
            install(ContentNegotiation) { json(json) }
        }
    } else {
        null
    }

    val isConfigured: Boolean
        get() = client != null

    suspend fun registerUser(nickname: String, pin: String): AuthSession =
        performAuthRpc("register_user", NicknameParams(pNickname = nickname, pPin = pin))

    suspend fun loginUser(nickname: String, pin: String): AuthSession =
        performAuthRpc("login_user", NicknameParams(pNickname = nickname, pPin = pin))

    suspend fun submitSession(nickname: String, durationSeconds: Int, sessionToken: String) {
        if (durationSeconds < AppConstants.MIN_SESSION_SECONDS) return
        executeVoidRpc(
            "submit_session",
            SessionParams(
                pNickname = nickname,
                pDuration = durationSeconds,
                pSessionToken = sessionToken,
            ),
        )
    }

    suspend fun fetchUserStats(nickname: String): UserStats {
        val http = client ?: throw SupabaseServiceError.NotConfigured
        val normalizedNickname = NicknameValidator.normalized(nickname)

        runCatching {
            val stats: List<UserStats> = fetchRpc(
                "get_user_stats",
                UserStatsParams(pNickname = normalizedNickname),
            )
            stats.firstOrNull()?.let { return it }
        }

        val stats: List<UserStats> = http.get("$baseUrl/rest/v1/user_stats") {
            applyHeaders()
            parameter("select", "total_duration_seconds,session_count")
            parameter("nickname", "ilike.$normalizedNickname")
            parameter("limit", 1)
        }.body()

        return stats.firstOrNull() ?: throw SupabaseServiceError.NetworkFailure
    }

    suspend fun fetchLeaderboard(limit: Int = AppConstants.LEADERBOARD_LIMIT): List<LeaderboardEntry> {
        val http = client ?: throw SupabaseServiceError.NotConfigured
        return http.get("$baseUrl/rest/v1/records") {
            applyHeaders()
            parameter("select", "*")
            parameter("order", "duration_seconds.desc")
            parameter("limit", limit)
        }.body()
    }

    suspend fun fetchTotalLeaderboard(limit: Int = AppConstants.LEADERBOARD_LIMIT): List<TotalLeaderboardEntry> {
        val http = client ?: throw SupabaseServiceError.NotConfigured
        return http.get("$baseUrl/rest/v1/user_stats") {
            applyHeaders()
            parameter("select", "nickname,total_duration_seconds")
            parameter("total_duration_seconds", "gt.0")
            parameter("order", "total_duration_seconds.desc")
            parameter("limit", limit)
        }.body()
    }

    private suspend fun performAuthRpc(function: String, params: NicknameParams): AuthSession {
        return try {
            decodeAuthSession(function, params)
        } catch (error: SupabaseServiceError) {
            throw error
        } catch (error: Exception) {
            throw mapUnexpectedAuthError(error)
        }
    }

    private suspend fun decodeAuthSession(function: String, params: NicknameParams): AuthSession {
        val http = client ?: throw SupabaseServiceError.NotConfigured
        val response = http.post("$baseUrl/rest/v1/rpc/$function") {
            applyHeaders()
            contentType(ContentType.Application.Json)
            setBody(params)
        }
        val body = response.bodyAsText()
        if (!response.status.isSuccess()) {
            throw SupabaseRpcErrorMapper.mapMessage(body.lowercase())
        }
        return AuthResponseDecoder.decodeSession(body, params.pNickname)
    }

    private suspend fun executeVoidRpc(function: String, params: SessionParams, mapErrors: Boolean = true) {
        val http = client ?: throw SupabaseServiceError.NotConfigured
        try {
            val response = http.post("$baseUrl/rest/v1/rpc/$function") {
                applyHeaders()
                contentType(ContentType.Application.Json)
                setBody(params)
            }
            if (!response.status.isSuccess()) {
                val body = response.bodyAsText()
                throw if (mapErrors) {
                    SupabaseRpcErrorMapper.mapMessage(body.lowercase())
                } else {
                    SupabaseServiceError.NetworkFailure
                }
            }
        } catch (error: SupabaseServiceError) {
            throw error
        } catch (error: Exception) {
            throw if (mapErrors) SupabaseRpcErrorMapper.map(error) else SupabaseServiceError.NetworkFailure
        }
    }

    private suspend inline fun <reified T> fetchRpc(
        function: String,
        params: UserStatsParams,
        mapErrors: Boolean = true,
    ): T {
        val http = client ?: throw SupabaseServiceError.NotConfigured
        return try {
            val response = http.post("$baseUrl/rest/v1/rpc/$function") {
                applyHeaders()
                contentType(ContentType.Application.Json)
                setBody(params)
            }
            if (!response.status.isSuccess()) {
                val body = response.bodyAsText()
                throw if (mapErrors) {
                    SupabaseRpcErrorMapper.mapMessage(body.lowercase())
                } else {
                    SupabaseServiceError.NetworkFailure
                }
            }
            response.body()
        } catch (error: SupabaseServiceError) {
            throw error
        } catch (error: Exception) {
            throw if (mapErrors) SupabaseRpcErrorMapper.map(error) else SupabaseServiceError.NetworkFailure
        }
    }

    private fun mapUnexpectedAuthError(error: Exception): SupabaseServiceError {
        if (SupabaseRpcErrorMapper.isLikelyEmptyAuthResponse(error)) {
            return SupabaseServiceError.ServerUpgradeRequired
        }
        return SupabaseRpcErrorMapper.map(error)
    }

    private fun io.ktor.client.request.HttpRequestBuilder.applyHeaders() {
        val key = SupabaseConfig.anonKey ?: throw SupabaseServiceError.NotConfigured
        header("apikey", key)
        header("Authorization", "Bearer $key")
    }

    private val baseUrl: String
        get() = SupabaseConfig.url ?: throw SupabaseServiceError.NotConfigured

    companion object {
        val shared = SupabaseService()
    }
}
